`rustc_parse` top-level cleanups

nnethercote opened the pull request:
A bunch of improvements in and around `compiler/rustc_parse/src/lib.rs`. Many of the changes streamline the API in that file from this (12 functions and one macro):
```
    name                              args                  return type
    ----                              ----                  -----------
    panictry_buffer!                  Result<T, Vec<Diag>>  T

pub parse_crate_from_file             path                  PResult<Crate>
pub parse_crate_attrs_from_file       path                  PResult<AttrVec>
pub parse_crate_from_source_str       name,src              PResult<Crate>
pub parse_crate_attrs_from_source_str name,src              PResult<AttrVec>

pub new_parser_from_source_str        name,src              Parser
pub maybe_new_parser_from_source_str  name,src              Result<Parser, Vec<Diag>>
pub new_parser_from_file              path,error_sp         Parser
    maybe_source_file_to_parser       srcfile               Result<Parser, Vec<Diag>>

pub parse_stream_from_source_str      name,src,override_sp  TokenStream
pub source_file_to_stream             srcfile,override_sp   TokenStream
    maybe_file_to_stream              srcfile,override_sp   Result<TokenStream, Vec<Diag>>

pub stream_to_parser                  stream,subparser_name Parser
```
to this:
```
    name                              args                  return type
    ----                              ----                  -----------
    unwrap_or_emit_fatal              Result<T, Vec<Diag>>  T
  
pub new_parser_from_source_str        name,src              Result<Parser, Vec<Diag>>
pub new_parser_from_file              path,error_sp         Result<Parser, Vec<Diag>>
    new_parser_from_source_file       srcfile               Result<Parser, Vec<Diag>>

pub source_str_to_stream              name,src,override_sp  Result<TokenStream, Vec<Diag>>
    source_file_to_stream             srcfile,override_sp   Result<TokenStream, Vec<Diag>>
```
I found the old API quite confusing, with lots of similar-sounding function names and no clear structure. I think the new API is much better.

r? @spastorino

nnethercote committed:
`UNICODE_ARRAY` and `ASCII_ARRAY` fixes.

- Avoid unnecessary escaping of single quotes within string literals.
- Add a missing blank line between two `UNICODE_ARRAY` sections.

nnethercote committed:
Don't use the word "parse" for lexing operations.

Lexing converts source text into a token stream. Parsing converts a
token stream into AST fragments. This commit renames several lexing
operations that have "parse" in the name. I think these names have been
subtly confusing me for years.

This is just a `s/parse/lex/` on function names, with one exception:
`parse_stream_from_source_str` becomes `source_str_to_stream`, to make
it consistent with the existing `source_file_to_stream`. The commit also
moves that function's location in the file to be just above
`source_file_to_stream`.

The commit also cleans up a few comments along the way.

nnethercote committed:
Inline and remove `error_malformed_cfg_attr_missing`.

It has a single call site.

This also means `CFG_ATTR_{GRAMMAR_HELP,NOTE_REF}` can be moved into
`parse_cfg_attr`, now that it's the only function that uses them.
And the commit removes the line break in the URL.

nnethercote committed:
Rename `maybe_file_to_stream` as `maybe_source_file_to_stream`.

Because it takes an `Lrc<SourceFile>`, and for consistency with
`source_file_to_stream`.

nnethercote committed:
Remove `stream_to_parser`.

It's a zero-value wrapper of `Parser::new`.

nnethercote committed:
Improve `panictry_buffer!`.

- Convert it from a macro to a function, which is nicer.
- Rename it as `unwrap_or_emit_fatal`, which is clearer.
- Fix the comment. In particular, `panictry!` no longer exists.
- Remove the unnecessary `use` declaration.

nnethercote committed:
Remove low-value comments.

The first one is out-of-date -- there are no longer functions expr,
item, stmt. And I don't know what a "HOF" is.

The second one doesn't really tell you anything.

nnethercote committed:
Rename `maybe_source_file_to_parser` as `maybe_new_parser_from_source_file`.

For consistency with `new_parser_from_{file,source_str}` and
`maybe_new_parser_from_source_str`.

nnethercote committed:
Use `source_str_to_stream` in a test file.

It does exactly what is required.

nnethercote committed:
Inline and remove `parse_crate{,_attrs}_from_{file,source_str}`.

All four functions are simple and have a single call site.

This requires making `Parser::parse_inner_attributes` public, which is
no big deal.

nnethercote committed:
Reorder `source_str_to_stream` arguments.

It's the only one of these functions where `psess` isn't the first
argument.

nnethercote committed:
Inline and remove `source_file_to_stream`.

It has a single call site.

nnethercote committed:
Make top-level `rustc_parse` functions fallible.

Currently we have an awkward mix of fallible and infallible functions:
```
       new_parser_from_source_str
 maybe_new_parser_from_source_str
       new_parser_from_file
(maybe_new_parser_from_file)        // missing
      (new_parser_from_source_file) // missing
 maybe_new_parser_from_source_file
       source_str_to_stream
 maybe_source_file_to_stream
```
We could add the two missing functions, but instead this commit removes
of all the infallible ones and renames the fallible ones leaving us with
these which are all fallible:
```
new_parser_from_source_str
new_parser_from_file
new_parser_from_source_file
source_str_to_stream
source_file_to_stream
```
This requires making `unwrap_or_emit_fatal` public so callers of
formerly infallible functions can still work.

This does make some of the call sites slightly more verbose, but I think
it's worth it for the simpler API. Also, there are two `catch_unwind`
calls and one `catch_fatal_errors` call in this diff that become
removable thanks this change. (I will do that in a follow-up PR.)

nnethercote commented:
Best reviewed one commit at a time. The commit messages explain each commit.

nnethercote commented on src/tools/rustfmt/src/parse/parser.rs:
```diff
@@ -68,10 +69,10 @@ impl<'a> ParserBuilder<'a> {
     ) -> Result<rustc_parse::parser::Parser<'a>, Option<Vec<Diag<'a>>>> {
         match input {
             Input::File(ref file) => catch_unwind(AssertUnwindSafe(move || {
-                new_parser_from_file(psess, file, None)
+                unwrap_or_emit_fatal(new_parser_from_file(psess, file, None))
```
Correct. `unwrap_or_emit_fatal(new_parser_from_file(...))` is equivalent to the old `new_parser_from_file(...)`. So there is no change to existing behaviour.

A nice thing about this is that the `catch_unwind` will be removable in a follow-up, removing the asymmetry between the `Input::File` and `Input::Text` cases here.

ytmimi replied:
That's great! I appreciate the explanation.

nnethercote replied:
I've had time to do the follow-up, which I added as an extra commit to this PR.

ytmimi replied:
Awesome! Thanks for the updated :)

nnethercote committed:
rustfmt: Remove an unnecessary `catch_unwind` use.

The `Input::File` and `Input::Text` cases should be very similar.
However, currently the `Input::File` case uses `catch_unwind` because,
until recently (#125815) there was a fallible version of
`new_parser_from_source_str` but only an infallible version of
`new_parser_from_file`. This difference wasn't fundamental, just an
overlooked gap in the API of `rustc_parse`.

Both of those operations are now fallible, so the `Input::File` and
`Input::Text` cases can made more similar, with no need for
`catch_unwind`. This also lets us simplify an `Option<Vec<Diag>>` to
`Vec<Diag>`.

spastorino commented:
@bors r+ rollup

bors commented:
:umbrella: The latest upstream changes (presumably #125989) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote commented:
I rebased.

@bors r=spastorino

bors commented:
:pushpin: Commit 2d4e7df14422ec46422e79ddae5af3ed78632bef has been approved by `spastorino`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=spastorino 2d4e7df14422ec46422e79ddae5af3ed78632bef -->
<!-- homu: {"type":"Approved","sha":"2d4e7df14422ec46422e79ddae5af3ed78632bef","approver":"spastorino","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
